tdesktop: fix calls and voice chat and clean up tg_owt

oxalica opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Fix #126285. 
According to @ilya-fedin https://github.com/NixOS/nixpkgs/issues/126285#issuecomment-882971211, the issue is caused by `libtgvoip` accidentally use another copy of webrtc library instead the `tg_owt` in `tdesktop`.

The solution should be:
1. Make `libtgvoip` also link to `tg_owt`.
2. Remove `libtgvoip` and use bundled one, which links to `tg_owt`.

I choose solution 2 because:
1. It's an internal library which is only used by telegram itself, and it's fine to use its bundled source in telegram.
2. It's forced to produce static library, is very fast to build (less than 1min) and is quite small (<2 MiB).
3. Reduce extra maintenance work and have almost no other drawback.


Things done:
- Remove system package `libtgvoip` and use bundled one for `tdesktop`.
- Do some clean up for `tg_owt` ~~and make it a shared library (like how Arch did)~~ (Arch also builds static library).

Current status:
- Voice message works.
- Voice call works.
- Group voice chat works.
- When no microphone is available, clicking `Setting` during voice chat causes the whole program to freeze.
  But It should be another unrelated issue.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

r? @primeos

oxalica committed:
tdesktop.tg_owt: rework and clean up

- Build shared library to keep binary small.

- Prefer system-wide libraries instead of bundled ones.

- `abseil-cpp` (absl) is required for downstream build and appears in
  output cmake file. So it's added to `propagatedBuildInputs`.

oxalica committed:
tdesktop: use bundled libtgvoip

oxalica committed:
libtgvoip: remove

It's removed because:
1. It's an internal library which is only used by telegram itself, and
   it's fine to use its bundled source in telegram.
2. It's forced to produce static library, is very fast to build
   (less than 1min) and is quite small (<2 MiB).
3. Reduce extra maintenance work and have almost no other drawback.

FliegendeWurst commented:
Voice messages and calls also work on 21.05 after backporting this change.

ilya-fedin commented on pkgs/applications/networking/instant-messengers/telegram/tdesktop/tg_owt.nix:
```diff
@@ -24,15 +24,15 @@ stdenv.mkDerivation {
 
-  cmakeFlags = [
-    # Building as a shared library isn't officially supported and currently broken:
-    "-DBUILD_SHARED_LIBS=OFF"
```
That's not a good idea, tg_owt has no stable ABI and share build is community supported (breaks after each big update and requires a long time to be fixed)

oxalica replied:
Fixed.

ilya-fedin commented:
> Do some clean up for tg_owt and make it a shared library (like how Arch did).

![изображение](https://user-images.githubusercontent.com/17829319/126443225-6b1ad6d3-1bae-498d-9507-e55eb9b5eebc.png)

r-rmcgibbo commented:
Result of `nixpkgs-review pr 130827` at 1b567a2c run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>tdesktop</li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl3` is a deprecated license, please check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/applications/networking/instant-messengers/telegram/tdesktop/default.nix:115:5:
```
    |
115 |     license = licenses.gpl3;
    |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

util-linux is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/applications/networking/instant-messengers/telegram/tdesktop/default.nix:57:3:
```
   |
57 |   buildInputs = [
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

extra-cmake-modules is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/applications/networking/instant-messengers/telegram/tdesktop/default.nix:57:3:
```
   |
57 |   buildInputs = [
   |   ^
```
</li>
  </ul>
</details>

oxalica committed:
tdesktop: clean up and fix outdated comments

ilya-fedin commented on pkgs/applications/networking/instant-messengers/telegram/tdesktop/default.nix:
```diff
@@ -52,16 +52,20 @@ in mkDerivation rec {
 
-  nativeBuildInputs = [ pkg-config cmake ninja python3 wrapGAppsHook wrapQtAppsHook removeReferencesTo ];
+  nativeBuildInputs = [
+    pkg-config cmake ninja python3 wrapGAppsHook wrapQtAppsHook removeReferencesTo
```
removeReferencesTo doesn't seem to be used

oxalica replied:
Removed.

ilya-fedin commented on pkgs/applications/networking/instant-messengers/telegram/tdesktop/default.nix:
```diff
@@ -70,7 +74,6 @@ in mkDerivation rec {
     "-DTDESKTOP_API_ID=611335"
     "-DTDESKTOP_API_HASH=d524b414d21f4d37f08684c1df41ac9c"
-    #"-DDESKTOP_APP_SPECIAL_TARGET=\"\"" # TODO: Error when set to "": Bad special target '""'
     "-DTDESKTOP_LAUNCHER_BASENAME=telegramdesktop" # Note: This is the default
```
Maybe drop that line as well? It's up to upstream how to name the desktop file, isn't it? I think the main reason this option exist are flatpak and snap packages.

oxalica replied:
Removed. Seems `telegramdesktop` is just the default value.

ilya-fedin commented:
Maybe it's worth to update the comment about rlottie/libtgvoip/mapbox-variant? mapbox-variant is not used for a long time, bundled rlottie is patched, non-bundled libtgvoip doesn't update its build scripts to the actual state. `-DTDESKTOP_USE_PACKAGED_TGVOIP=ON` and `-DDESKTOP_APP_USE_PACKAGED_RLOTTIE=ON` don't exist for a long time, tdesktop always uses autodetect and fallbacks to bundled copies.

ilya-fedin commented:
I also wonder whether libtgvoip's dlopen calls work or not:
https://github.com/telegramdesktop/libtgvoip/blob/tdesktop/os/linux/AudioInputALSA.cpp#L26
https://github.com/telegramdesktop/libtgvoip/blob/tdesktop/os/linux/AudioOutputALSA.cpp#L25
https://github.com/telegramdesktop/libtgvoip/blob/tdesktop/os/linux/AudioPulse.cpp#L73

oxalica commented:
> I also wonder whether libtgvoip's dlopen calls work or not

I'm not familiar with the code, but when would they be called? How can I test it?
I tried some voice chat but seems no error log related to `asound` or `libpulse` is printed.

ilya-fedin commented:
> but when would they be called?

When you making a call to an old client (pre-video calls version)

oxalica commented:
I added `alsa-lib` and `libpulseaudio` to `LD_LIBRARY_PATH`. `dlopen` should work now.
But it's hard for me to test running related code.

ilya-fedin commented:
> I added `alsa-lib` and `libpulseaudio` to `LD_LIBRARY_PATH`.

Isn't it better to use substituteInPlace?
```
  postPatch = ''
    substituteInPlace Telegram/ThirdParty/libtgvoip/os/linux/AudioInputALSA.cpp \
      --replace '"libasound.so.2"' '"${alsa-lib}/lib/libasound.so.2"'
    substituteInPlace Telegram/ThirdParty/libtgvoip/os/linux/AudioOutputALSA.cpp \
      --replace '"libasound.so.2"' '"${alsa-lib}/lib/libasound.so.2"'
    substituteInPlace Telegram/ThirdParty/libtgvoip/os/linux/AudioPulse.cpp \
      --replace '"libpulse.so.0"' '"${libpulseaudio}/lib/libpulse.so.0"'
  '';
```

ilya-fedin commented:
IMHO it's worth to think about setting `-DDESKTOP_APP_USE_PACKAGED_FONTS=OFF` since it's impossible to install fonts as dependencies of packages with Nix and tdesktop's widgets are developed only with Open Sans in mind (it has a lot of hardcoded values and wide fonts like DejaVu may even go out of widgets' bounds)

oxalica commented:
> Isn't it better to use substituteInPlace?

There are already paths scattered (and hard-coded) in three files. They may add another `dlopen` somewhere else at anytime.

I found in [nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#setup-hook-autopatchelfhook) that there is a `runtimeDependencies` which would be added into RPATH by patchelf. Would it be better?
(Not sure how it behaves on darwin)

ilya-fedin commented:
> There are already paths scattered (and hard-coded) in three files. They may add another `dlopen` somewhere else at anytime.

libtgvoip is a legacy library, I don't think that would happen.

> I found in [nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#setup-hook-autopatchelfhook) that there is a `runtimeDependencies` which would be added into RPATH by patchelf. Would it be better?
> (Not sure how it behaves on darwin)

It's a part of autoPatchElfHook, for prebuilt binaries.

oxalica commented:
> It's worth to think about setting -DDESKTOP_APP_USE_PACKAGED_FONTS=OFF

Currently the option is `ON` and it can use system font I set in `configuration.nix`. If we turn off the option, does it mean it will no longer use system fonts, and instead always use embedded fonts? That may not be acceptable since I need also Japanese and Chinese fonts.

ilya-fedin commented:
> Currently the option is `ON` and it can use system font I set in `configuration.nix`. If we turn off the option, does it mean it will no longer use system fonts, and instead always use embedded fonts? That may not be acceptable since I need also Japanese and Chinese fonts.

The options is never supposed to provide ability to use system fonts. If you use it, you must provide open sans with tdesktop in some way. If you can't provide it, let it use embedded copy of open sans. Or the UI will be broken.

This has nothing to do with Japanese or Chinese, though.

oxalica committed:
tdesktop: patch dlopen paths

oxalica committed:
tdesktop: bundle the default font

It's worth to think about setting -DDESKTOP_APP_USE_PACKAGED_FONTS=OFF
since it's impossible to install fonts as dependencies of packages with
Nix and tdesktop's widgets are developed only with Open Sans in mind (it
has a lot of hardcoded values and wide fonts like DejaVu may
even go out of widgets' bounds)

https://github.com/NixOS/nixpkgs/pull/130827#issuecomment-885212649

oxalica commented:
Rebased and squashed some commits

primeos commented on pkgs/applications/networking/instant-messengers/telegram/tdesktop/default.nix:
```diff
@@ -113,10 +122,10 @@ in mkDerivation rec {
     homepage = "https://desktop.telegram.org/";
-    changelog = "https://github.com/telegramdesktop/tdesktop/releases/tag/v{version}";
-    maintainers = with maintainers; [ primeos abbradar ];
+    changelog = "https://github.com/telegramdesktop/tdesktop/releases/tag/v${version}";
```
Thanks for fixing that embarrassing mistake btw :) Seems like I never tested/noticed it... :o

primeos reviewed (approved):
Diff LGTM, huge thanks for the countless improvements! :)

(Only nitpick: The first commit message is outdated and the diff breaks the ordering a bit but that doesn't seem important enough to warrant an additional iteration.)

primeos commented:
I did some very basic testing and didn't notice any regressions.
Thanks again!

drupol commented:
nice ! Would be very nice to have in `release-21.05` !

kvtb commented:
>  If you can't provide it, let it use embedded copy of open sans. Or the UI will be broken.

What do you mean as "UI broken"?
I tried with system fonts and UI seems good (I do not like Open Sans, also it makes `tdesktop` look too alien among other Linux apps)

kvtb commented on pkgs/applications/networking/instant-messengers/telegram/tdesktop/default.nix:
```diff
@@ -46,22 +46,33 @@ in mkDerivation rec {
+  ];
 
   buildInputs = [
     qtbase qtimageformats gtk3 libsForQt5.kwayland libsForQt5.libdbusmenu lz4 xxHash
```
```diff
-  qtbase qtimageformats gtk3 libsForQt5.kwayland libsForQt5.libdbusmenu lz4 xxHash
+  qtbase qtimageformats gtk3 kwayland libdbusmenu lz4 xxHash
```

`kwayland` and `libdbusmenu` are in the scope, the file is called using
`qt5.callPackage ../applications/networking/instant-messengers/telegram/tdesktop { }`

oxalica replied:
Can you open a new PR to fix them? Does it causing any issues?

kvtb replied:
It resulted in two places where Qt's version is specified, independently:

I backported new `tdesktop` to older NixOS, when default QT is `5.12`
Changing `qt5.callPackage` to `qt515.callPackage` was not enough to chose `5.15`, because these `libsForQt5` still pointed to default `pkgs.qt5` (which is 5.12).

As `tdesktop` tends to use bleeding edge versions of C++, Qt and other deps, eventually someone else will hit into the problem compiling it together with two different `Qt`s

SuperSandro2000 replied:
Doing a PR with the fixes.

primeos replied:
#132860

kvtb commented on pkgs/applications/networking/instant-messengers/telegram/tdesktop/default.nix:
```diff
@@ -1,11 +1,11 @@
 { mkDerivation, lib, fetchFromGitHub, callPackage, fetchpatch
-, pkg-config, cmake, ninja, python3, wrapGAppsHook, wrapQtAppsHook, removeReferencesTo
+, pkg-config, cmake, ninja, python3, wrapGAppsHook, wrapQtAppsHook
 , qtbase, qtimageformats, gtk3, libsForQt5, lz4, xxHash
```
```diff
- , qtbase, qtimageformats, gtk3, libsForQt5, lz4, xxHash
+ , qtbase, qtimageformats, gtk3, kwayland, libdbusmenu, lz4, xxHash
```

ilya-fedin commented:
> What do you mean as "UI broken"?

![изображение](https://user-images.githubusercontent.com/17829319/128446809-8a70d4e4-c4a2-4fef-a226-d8847655ef3b.png)
(it's somewhere from tdesktop issues)
